system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,517 @@
+  return [firstLog, secondLog];
+}
+
+// bytecode
+interface BytecodeData {
+  content: string;
+  lengthBytes: string;
+}
+
+async function setupBytecodeData(l1MessengerAddress: string): Promise<BytecodeData> {
+  const content = await getCode(l1MessengerAddress);
+  const lengthBytes = ethers.utils.hexZeroPad(ethers.utils.hexlify(ethers.utils.arrayify(content).length), 4);
+  return {
+    content,
+    lengthBytes,
+  };
+}
+
+interface DataPair {
+  lengthBytes: string;
+  content: string;
+}
+
+function createTotalL2ToL1PubdataAndStateDiffs(
+  numberOfLogs: number,
+  logs: string[],
+  numberOfMessages: number,
+  messages: DataPair[],
+  numberOfBytecodes: number,
+  bytecodes: DataPair[],
+  compressedStateDiffsSizeBytes: string,
+  enumerationIndexSizeBytes: string,
+  compressedStateDiffs: string,
+  numberOfStateDiffsBytes: string,
+  encodedStateDiffs: string,
+  version: string = ethers.utils.hexZeroPad(ethers.utils.hexlify(1), 1)
+): string {
+  const messagePairs = [];
+  for (let i = 0; i < numberOfMessages; i++) {
+    messagePairs.push(messages[i].lengthBytes, messages[i].content);
+  }
+
+  const bytecodePairs = [];
+  for (let i = 0; i < numberOfBytecodes; i++) {
+    bytecodePairs.push(bytecodes[i].lengthBytes, bytecodes[i].content);
+  }
+
+  return ethers.utils.concat([
+    ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfLogs), 4),
+    ...logs,
+    ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfMessages), 4),
+    ...messagePairs,
+    ethers.utils.hexZeroPad(ethers.utils.hexlify(numberOfBytecodes), 4),
+    ...bytecodePairs,
+    version,
+    compressedStateDiffsSizeBytes,
+    enumerationIndexSizeBytes,
+    compressedStateDiffs,
+    numberOfStateDiffsBytes,
+    encodedStateDiffs,
+  ]);
+}
+class L1MessengerPubdataEmulator {
+  numberOfLogs: number;
+  logs: string[];
+  numberOfMessages: number;
+  messages: DataPair[];
+  numberOfBytecodes: number;
+  bytecodes: DataPair[];
+  stateDiffsSetupData: StateDiffSetupData;
+  version: string;
+
+  constructor() {
+    this.numberOfLogs = 0;
+    this.logs = [];
+    this.numberOfMessages = 0;
+    this.messages = [];
+    this.numberOfBytecodes = 0;
+    this.bytecodes = [];
+    this.stateDiffsSetupData = {
+      compressedStateDiffsSizeBytes: "",
+      enumerationIndexSizeBytes: "",
+      compressedStateDiffs: "",
+      numberOfStateDiffsBytes: "",
+      encodedStateDiffs: "",
+    };
+    this.version = ethers.utils.hexZeroPad(ethers.utils.hexlify(1), 1);
+  }
+
+  addLog(log: string) {
+    this.logs.push(log);
+    this.numberOfLogs++;
+  }
+
+  addMessage(message: DataPair) {
+    this.messages.push(message);
+    this.numberOfMessages++;
+  }
+
+  addBytecode(bytecode: DataPair) {
+    this.bytecodes.push(bytecode);
+    this.numberOfBytecodes++;
+  }
+
+  setStateDiffsSetupData(data: StateDiffSetupData) {
+    this.stateDiffsSetupData = data;
+  }
+
+  buildTotalL2ToL1PubdataAndStateDiffs() {
+    return createTotalL2ToL1PubdataAndStateDiffs(
```

Comment:
Taking into account that this is the only place where the function is invoked, do you think it would make sense to actually move the function here? The function `createTotalL2ToL1PubdataAndStateDiffs` takes a lot of parameters and the amount of code could be reduced by accessing those variables with `this`, but this is opinionated

Reply:
fixed